tests/GcodeLayerThreaderTest.cpp

```diff
@@ -0,0 +1,123 @@
+//Copyright (c) 2015 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "GcodeLayerThreaderTest.h"
+
+#include <iomanip>
+#include <sstream> // ostringstream
+#include <list>
+
+#include <../src/utils/intpoint.h>
+#include <../src/GcodeLayerThreader.h>
+
+
+
+namespace cura
+{
+    CPPUNIT_TEST_SUITE_REGISTRATION(GcodeLayerThreaderTest);
+
+void GcodeLayerThreaderTest::setUp()
+{
+    //Do nothing.
+}
+
+void GcodeLayerThreaderTest::tearDown()
+{
+    //Do nothing.
+}
+
+void GcodeLayerThreaderTest::test1()
+{
+    test(1000, 100, 10);
+}
+void GcodeLayerThreaderTest::test2()
+{
+    test(100, 1000, 10);
+}
+void GcodeLayerThreaderTest::test3()
+{
+    test(1000, 20, 1);
+}
+void GcodeLayerThreaderTest::test4()
+{
+    test(10000, 1000, 100);
+}
+
+
+
+void GcodeLayerThreaderTest::test(int avg_computation_time, int layer_count, int max_task_count)
+{
+//     srand(1);
+    int starting_layer_nr = -10;
+    
+    std::vector<int> times_consumed;
+    std::vector<int> times_produced;
+    times_consumed.resize(layer_count, 0);
+    times_produced.resize(layer_count, 0);
+    
+    const std::function<LayerPlan* (int)>& produce_item =
+        [&times_produced, &starting_layer_nr, &avg_computation_time](int layer_nr)
+        {
+            // simulate work:
+            std::this_thread::sleep_for(std::chrono::milliseconds(rand() % (avg_computation_time * 2 * 2 / 3)));
```

Comment:
Using sleep in a unit test is very bad practice! And `rand()` too.

This unit test is now nondeterministic. And many many times slower.